feat(profileNote): input field and allow user to store private note

nickjjang opened the pull request:
**What this PR does** 📖
Profile > Note -> Add 'Notes' input field and allow user to store a private note about that user/edit note

**Which issue(s) this PR fixes** 🔨
AP-519

**Special notes for reviewers** 🗒️
 - Build the click to edit UI components in /components/views/user/profile/ClickToEdit
 - To write the note into textile, make the MetadataManager in /libraries/Textile/MetadataManager

**Additional comments** 🎤
![image](https://user-images.githubusercontent.com/81814644/156534571-c36a810f-2e8f-475c-b600-1629feb84865.png)

WanderingHogan commented on components/views/user/profile/ClickToEdit/ClickToEdit.vue:
```diff
@@ -0,0 +1,60 @@
+      this.html = value
+      if (this.focused) {
+        el.innerHTML = value
+      } else {
```
Can you do a guard here

WanderingHogan commented on components/views/user/profile/ClickToEdit/ClickToEdit.vue:
```diff
@@ -0,0 +1,60 @@
+  props: {
+    placeholder: {
+      type: String,
+      default: 'Click to add note',
```
Can you add this to translations

WanderingHogan commented on libraries/Textile/MetadataManager.ts:
```diff
@@ -0,0 +1,162 @@
+    }
+    return this.threadID
+  }
+
```
Can you add the basic docs to these methods

WanderingHogan commented on libraries/Textile/MetadataManager.ts:
```diff
@@ -0,0 +1,162 @@
+    if (record) {
+      record.body = body
+      await this.textile.client.save(threadID, CollectionName, [record])
+    } else {
```
can you do a guard here

WanderingHogan reviewed (changes requested):
Hey - this is so exciting and neat!

Couple requests:

- Move `components/views/user/profile/ClickToEdit` to `components/interactables/ClickToEdit` make anything that prevents it from being generic a prop, i didn't notice anything when i was looking at it
- Address Phill's comments with when the user hits enter
- Make some minor updates (commented on this ticket)

nickjjang commented:
Hi @WanderingHogan, @phillsatellite I've fixed issues you addressed. Do you have time to review it again. Thank you.
@iltumio, @WanderingHogan, I've added MetadataManager in libraries/Textile/MetadataManager to add note in friend's profile. but it can be used to store user's metadata information.
What do you think about it? Is it okay to manage data from MetadataManager? or do you have another good idea?

josephmcg reviewed (commented):
minor code cleanup

molimauro commented on components/views/user/profile/About.vue:
```diff
@@ -0,0 +1,68 @@
+      set(note: string) {
+        const { userProfile } = this.ui
+        let { metadata } = userProfile
+        metadata = !metadata
+          ? {
+              note,
+            }
+          : {
+              ...metadata,
+              note,
+            }
```
```
{
   ...(metadata ?? null)
   note,
}
```

iltumio commented on components/interactables/ClickToEdit/ClickToEdit.html:
```diff
@@ -0,0 +1,11 @@
+<div class="cte-container">
+  <div
```
can we use the Editable component introduced by Mauro for the chatbar instead?

iltumio commented on libraries/Textile/MetadataManager.ts:
```diff
@@ -0,0 +1,200 @@
+   * @returns a string name of the created metadata collection
+   */
+  async getMetadataCollection(): Promise<string> {
+    const thread = await this.textile.users.getThread('hubmail')
```
we can probably use a different thread name since this one is related to the mailbox api

nickjjang replied:
when I tested thread is database and I couldn't get other database. textile says database is not created.

iltumio commented on libraries/Textile/MetadataManager.ts:
```diff
@@ -0,0 +1,200 @@
+  textile: TextileInitializationData
+  threadID?: ThreadID
+
+  constructor(textile: TextileInitializationData, senderAddress: string) {
```
why do we need to pass the `senderAddress` since it is already present in `TextileInitializationData`?

iltumio commented on libraries/Textile/TextileManager.ts:
```diff
@@ -83,6 +85,10 @@ export default class TextileManager {
     )
+
+    // MetadataManager initializes itself during the creation
+    this.metadataManager = new MetadataManager(textile, textile.wallet.address)
```
see my previous comment

iltumio commented on libraries/Textile/schema.ts:
```diff
@@ -13,3 +13,18 @@ export const groupChatSchema = {
+
+export const metadataSchema = {
+  $schema: 'http://json-schema.org/draft-07/schema#',
+  title: 'Message',
```
wrong title. Shouldn't it be `Note`?

iltumio commented on libraries/Textile/schema.ts:
```diff
@@ -13,3 +13,18 @@ export const groupChatSchema = {
+    _id: { type: 'string' },
+    created_at: { type: 'number' },
+    key: { type: 'string' },
+    from: { type: 'string' },
```
do we need `from` and `to`?

iltumio commented on store/friends/actions.ts:
```diff
@@ -171,6 +187,41 @@ export default {
+      throw new Error(TextileError.METADATA_MANAGER_NOT_FOUND)
+    }
+    const $MetadataManager: MetadataManager = $TextileManager.metadataManager
+    friend.metadata = metadata
```
don't understand this line

iltumio commented on types/textile/metadata.ts:
```diff
@@ -0,0 +1,10 @@
+
+export interface MetadataFromThread {
+  _id: string
+  from: string
```
do we really need `from` and `to`?

nickjjang commented on components/views/user/profile/About.vue:
```diff
@@ -0,0 +1,68 @@
+  computed: {
+    ...mapState(['friends', 'ui']),
+    note: {
+      get(): string {
+        if (
+          this.ui.userProfile &&
+          this.ui.userProfile.metadata &&
+          this.ui.userProfile.metadata.note
+        )
+          return this.ui.userProfile.metadata.note
+        return ''
```
noted me thanks

phillsatellite commented:
Tested again: Found one other small issue, if a user adds a lot of text the input box will extend and text will go past the edge of Profile Modal 
https://user-images.githubusercontent.com/93608357/157719568-6099fec5-01ee-4fcb-8155-7e09e2d0dc80.mov

phillsatellite commented:
Tested with changes made: kinda strange was still able to replicate previous bug

https://user-images.githubusercontent.com/93608357/157964441-44629228-543f-44fa-8a0a-4f11bb59c4c3.mov

nickjjang commented:
> Tested with changes made: kinda strange was still able to replicate previous bug
> 
>  samebug.mov

Hi @phillsatellite
I applogize for the inconvenience. I've fixed and tested in chrome, and firefox. let me know if it works well on your side.

phillsatellite commented:
Tested: works a lot better! One last thing I checked with Design about, if the user saves a big note like in screenshot, they should be able to scroll through
<img width="1037" alt="Screen Shot 2022-03-14 at 12 01 03 PM" src="https://user-images.githubusercontent.com/93608357/158216355-eae5a609-2218-4b8f-accb-083033bebbd7.png">

figma 

<img width="1037" alt="Screen_Shot_2022-03-14_at_11 16 46_AM" src="https://user-images.githubusercontent.com/93608357/158216399-69729381-0473-44aa-8633-8e159acd3c70.png">

nickjjang commented:
Hi @phillsatellite 
Added scrollbar to the profile tab contents.

nickjjang commented on components/interactables/ClickToEdit/ClickToEdit.html:
```diff
@@ -0,0 +1,11 @@
+<div class="cte-container">
+  <div
```
![unknown](https://user-images.githubusercontent.com/81814644/158996004-e908493e-7069-4c90-aa7b-e6f5e4cfa5aa.png)
I cannot use views/chatbar/Editable.vue. if I use this component and type into Add Note  (red part). focus on Chatbar(yello part)  automatically and insert the text.

stavares843 commented:
<img width="1372" alt="Captura de ecrã 2022-03-18, às 17 04 05" src="https://user-images.githubusercontent.com/29093946/159049571-11d8f6a9-5974-47e9-ba69-e3829566b324.png">
<img width="1359" alt="Captura de ecrã 2022-03-18, às 17 04 16" src="https://user-images.githubusercontent.com/29093946/159049588-ad7ea105-668b-4652-b53a-379096702aea.png">


dev vs this branch

the blue tab was changed, it should be as dev, to be similar to the design

stavares843 commented:
<img width="841" alt="Captura de ecrã 2022-03-18, às 17 07 06" src="https://user-images.githubusercontent.com/29093946/159050703-67663877-1a2b-4f09-858c-abc782c0a2f8.png">

user is able to save a note with just spacing

stavares843 commented:
https://user-images.githubusercontent.com/29093946/159050759-9a9b07b6-853c-4915-b127-48a67ccb0f92.mov

when adding a huge note, sometimes there's no scroll bar and we need to scroll manually until the bottom

also, doesn't seem there's a limit on the chars, and browsers gets pretty slow adding a bunch of them, so we should add a limit on this

stavares843 commented:
https://user-images.githubusercontent.com/29093946/159050912-6b40de18-39c6-4b97-b5a2-e04c97354f27.mov

there's a glitch when you add a bunch of chars, select and then deleted them all, there's a glitch with the scroll bar

stavares843 commented:
https://user-images.githubusercontent.com/29093946/159051324-9184fb7b-adf0-4b44-a8f8-b319b7020015.mov

if you copy paste a whole web page, layout breaks

WanderingHogan commented:
@nickjjang great job on everything and thank you for working with us while we go through all this neat feature. @stavares843: Great finds, your eye for detail is amazing.

For now though, let's focus less on pixel perfect in our current timeframe right now trying to get to early access, though. So in review, Sara needs:

1. Fix the border on the tabs. I believe it's `components/views/user/profile/Profile.less` line 43, 'margin: 0 @large-spacing @large-spacing @large-spacing;' instead of padding should do it.

2. Prevent user from storing a blank or empty note, check out jeffs change in `components/views/chat/chatbar/Chatbar.vue` line 252 or your update to the shared input component

3. Cap the note length at 256, add this to config.ts just like we have the limit on chat input. This should also take care of the scroll bar glitch/prevent it from being a problem and the layout breaks

WanderingHogan commented:
Lets move these changes to another ticket

stavares843 commented:
I can add a ticket

stavares843 commented:
[AP-1142](https://satellite-im.atlassian.net/browse/AP-1142)
